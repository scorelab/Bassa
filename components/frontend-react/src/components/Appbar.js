import React, { useState } from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

// MUI imports
import { makeStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import Avatar from '@material-ui/core/Avatar';
import TextField from '@material-ui/core/TextField';
import InputAdornment from '@material-ui/core/InputAdornment';
import IconButton from '@material-ui/core/IconButton';
import Visibility from '@material-ui/icons/Visibility';
import VisibilityOff from '@material-ui/icons/VisibilityOff';
import { logoutUser } from '../actions/userActions';

// Logo import
import logo from '../logo.png';

const colours = {
  somewhat_brown: '#625042',
  white: '#fff'
};

const useStyles = makeStyles(theme => ({
  root: {
    flexGrow: 1
  },
  grow: {
    flexGrow: 1
  },
  menuButton: {
    margin: theme.spacing(2),
    marginTop: theme.spacing(3)
  },
  avatar: {
    margin: 10
  },
  bigAvatar: {
    margin: 10,
    width: 60,
    height: 60
  },
  close: {
    padding: theme.spacing(0.5)
  }
}));

const BassaAppBar = props => {
  const classes = useStyles();

  const { onClickLogin, isloggedIn } = props;

  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [showPassword, togglePasswordView] = useState(false);

  const handleClickShowPassword = () => {
    const value = showPassword;
    togglePasswordView(!value);
  };

  const handleMouseDownPassword = event => {
    event.preventDefault();
  };

  const handleUserField = event => {
    setUsername(event.target.value);
  };

  const handlePassField = event => {
    setPassword(event.target.value);
  };

  const handleLoginButton = event => {
    event.preventDefault();
    onClickLogin(username, password);
  };

  const handleNavigationButtons = route => {
    switch (route) {
      case 'logout':
        sessionStorage.removeItem('token');
        props.logoutUser();
        props.history.replace('/');
        break;

      case 'home':
        props.history.push('/home');
        break;

      case 'admin':
        props.history.push('/admin');
        break;

      default:
        break;
    }
  };

  if (!isloggedIn) {
    // If the user is not logged in, then return component for logged out user
    return (
      <div className={classes.root}>
        <AppBar
          position="static"
          style={{ background: colours.somewhat_brown }}
        >
          <Toolbar>
            <Avatar alt="Bassa" src={logo} className={classes.bigAvatar} />
            <Typography variant="h4" color="inherit" className={classes.grow}>
              Bassa
            </Typography>
            <form onSubmit={handleLoginButton}>
              <TextField
                id="username"
                data-test="input-username"
                margin="normal"
                variant="outlined"
                onChange={handleUserField}
                style={{ background: colours.white }}
                placeholder="Username"
              />
              &nbsp;&nbsp;
              <TextField
                id="password"
                data-test="input-password"
                type={showPassword ? 'text' : 'password'}
                variant="outlined"
                onChange={handlePassField}
                margin="normal"
                style={{ background: colours.white }}
                placeholder="Password"
                InputProps={{
                  endAdornment: (
                    <InputAdornment position="end">
                      <IconButton
                        edge="end"
                        aria-label="toggle password visibility"
                        onClick={handleClickShowPassword}
                        onMouseDown={handleMouseDownPassword}
                      >
                        {showPassword ? <VisibilityOff /> : <Visibility />}
                      </IconButton>
                    </InputAdornment>
                  )
                }}
              />
              &nbsp;&nbsp;
              <Button
                type="submit"
                className={classes.menuButton}
                size="large"
                variant="contained"
                color="primary"
              >
                Login
              </Button>
            </form>
          </Toolbar>
        </AppBar>
      </div>
    );
  }
  // Else return component for logged in user
  return (
    <div className={classes.root}>
      <AppBar position="static" style={{ background: colours.somewhat_brown }}>
        <Toolbar>
          <Avatar alt="Bassa" src={logo} className={classes.avatar} />
          <Typography variant="h6" color="inherit" className={classes.grow}>
            Bassa
          </Typography>
          <Button
            onClick={() => handleNavigationButtons('home')}
            data-test="button-dashboard"
            size="small"
            color="inherit"
          >
            Dashboard
          </Button>
          <Button
            onClick={() => handleNavigationButtons('admin')}
            data-test="button-admin"
            size="small"
            color="inherit"
          >
            Admin
          </Button>
          <Button
            onClick={() => handleNavigationButtons('logout')}
            data-test="button-logout"
            size="small"
            color="inherit"
          >
            Logout
          </Button>
        </Toolbar>
      </AppBar>
    </div>
  );
};

const mapDispatchToProps = dispatch => ({
  logoutUser: () => dispatch(logoutUser())
});

export default connect(
  null,
  mapDispatchToProps
)(withRouter(BassaAppBar));
