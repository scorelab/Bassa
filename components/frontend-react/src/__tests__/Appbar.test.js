/* eslint-disable no-return-assign */
/* eslint-disable no-undef */
import React from 'react';
import { shallow, mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import Appbar from '../components/Appbar';

const findByTestAttr = (component, attr) => {
  const wrapper = component.find(`[data-test='${attr}']`);
  return wrapper;
};

describe('Appbar Component Testing', () => {
  let component;
  beforeEach(() => (component = shallow(<Appbar />)));

  it('should match with its snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  describe('User is logged in', () => {
    beforeEach(
      () =>
        (component = mount(
          <BrowserRouter>
            <Appbar isloggedIn />
          </BrowserRouter>
        ))
    );

    it('should display logged in view', () => {
      const buttonDashboard = findByTestAttr(component, 'button-dashboard').at(
        4
      );
      const buttonAdmin = findByTestAttr(component, 'button-admin').at(4);
      const buttonLogout = findByTestAttr(component, 'button-logout').at(4);

      expect(buttonDashboard.props().children[0].props.children).toBe(
        'Dashboard'
      );
      expect(buttonAdmin.props().children[0].props.children).toBe('Admin');
      expect(buttonLogout.props().children[0].props.children).toBe('Logout');
    });
  });

  describe('User is logged out', () => {
    beforeEach(
      () =>
        (component = mount(
          <BrowserRouter>
            <Appbar isloggedIn={false} />
          </BrowserRouter>
        ))
    );

    it('should display logged out view', () => {
      const inputUsername = findByTestAttr(component, 'input-username').at(4);
      const inputPassword = findByTestAttr(component, 'input-password').at(4);

      expect(inputUsername.props().children[1].props.id).toBe('username');
      expect(inputPassword.props().children[1].props.id).toBe('password');
    });
  });
});
