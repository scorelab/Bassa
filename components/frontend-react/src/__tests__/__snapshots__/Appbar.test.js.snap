// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Appbar Component Testing should match with its snapshot 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
